src/luarocks/fetch.lua

```diff
@@ -355,7 +356,29 @@ function fetch.get_sources(rockspec, extract, dest_dir)
       if not ok then return nil, err end
       fs.unpack_archive(rockspec.source.file)
       if not fs.exists(rockspec.source.dir) then
-         return nil, "Directory "..rockspec.source.dir.." not found inside archive "..rockspec.source.file, "source.dir", source_file, store_dir
+
+         -- If rockspec.source.dir can't be found, see if we only have one
+         -- directory in store_dir.  If that's the case, assume it's what
+         -- we're looking for.
+         -- We only do this if the rockspec source.dir was not set, and only
+         -- with rockspecs newer than 3.0.
+         local dir_count, found_dir = 0
+
+         if not rockspec.source.dir_set and deps.format_is_at_least(rockspec, "3.0") then
+            local files = fs.list_dir()
+            for _, f in ipairs(files) do
```

Comment:
Oh, I see. Deleting it would probably be the best option, but a feature gets in the way: `source.url` can be a single *.lua or *.c, and `fs.unpack_archive` conveniently accepts *.lua and *.c files as a no-op, so again some logic to determine "this is really an archive file that can be deleted" would be needed, which brings us back to square one. :\

I think auto-detecting `"."` will have to stay out for now, then.